docs/design/direct-assign-volume.md

```diff
@@ -0,0 +1,139 @@
+on the host, and is more efficient from a storage IO perspective. 
+
+![Volume handling with direct-assignment](arch-images/DA-UC.png)
+
+# Proposal - Introduce pattern for directly assigning block device for mounting in the guest
+
+In the existing upstream, there is no direct communication between CSI and the container runtime. By the
+time volume information is presented to the runtime, it is a path on the given host filesystem. We need a method to:
+ - enlighten the CSI driver that a given PVC is to be utilized by Kata (or any other runtime that would prefer to mount the volume on their own).
+ - communication path between the CSI node driver and the runtime to detail: what the backing block device is,
+   what kind of filesystem it is, and where it should be mounted with respect to the container rootfs.
+
+Ideally we'd be able to come up with a solution that would facilitate a runtime calling into CSI instead of Kubelet, as is done today for CNI. Working within the existing constraints that exist, we are proposing a pattern that can be used to facilitate direct-volume assignment, while working with other upstream projects for a more direct method. 
+
+## Carrying out the mount
+
+The CSI node driver will need to be enlightened that the volume will be consumed by a VM-based runtime, and will need 
+to communicate with the runtime to describe where the block device lives, what filesystem is on it, and where it
+should be mounted within the container(s).
+
+### Specifiying that the volume will be consumed by a VM-friendly runtime
+
+If a persistent volume [(PV)](https://kubernetes.io/docs/concepts/storage/persistent-volumes/) is intended to be consumed by a VM-based runtime, the user can specify this by adding an annotation to their respect persistent volume claim (PVC). Example:
+
+```yaml
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  annotations:
+    skip-hostmount: "true"
+  name: amazing-kata-volume
+spec:
+  accessModes:
+    - ReadWriteOnce
+  resources:
+    requests:
+      storage: 4Gi
+```
+
+Based on this annotation, the CSI node driver would identify that the volume should not be mounted, and special handling is needed.
+
+### Communicating block device details to the runtime
+
+The runtime will need to identify that a mount is backed by a block device and that it needs to handle the direct-assignment use case.
+
+When a volume is presented for direct-assignment, the CSI driver will create a particular file, `csiPlugin.json`
+at the root of the volume on the host. The contents of this file will be defined as follows:
+
+```go
+type DiskMountInfo struct {
+        // Device: source device for this volume
+        Device string json:"device,omitempty"
+
+        // VolumeType: type associated with this volume (ie, block)
+        VolumeType string json:"volumeType,omitempty"
+
+        // TargetPath: path which this device should be mounted within the guest
+        TargetPath string json:"targetPath,omitempty"
```

Comment:
The implicit assumption in the entire proposal is that the volume will only be used by one pod on the node at a time.

The opposite is currently what Kubernetes does: a volume is mounted once and then shared by multiple pods, at least by CSI drivers which support NodeStageVolume.

https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/2485-read-write-once-pod-pv-access-mode will address that.

Reply:
This is a good point. One limitation is earlier, if there were two pods with affinity on the same node, they would be able to write to a block based PV in parallel (due to stage-once and bind-mount the global mount behavior). With the approach in this proposal, two pods with the affinity to the same node won't be able to write to a block based PV unless the FS can handle parallel rw mounts. This should be called out.

Other than the above scenario, this proposal does not prevent multiple Kata containers on the same (or different) nodes from accessing the same PV as long as the filesystem (on block or over network) can handle the parallel access. So it is not expected all CSI plugins utilizing this optimization will have to declare `SINGLE_NODE_SINGLE_WRITER`